Specialize `iter::ArrayChunks::fold` for TrustedRandomAccess iterators

the8472 opened the pull request:
```
OLD:
test iter::bench_trusted_random_access_chunks                      ... bench:         368 ns/iter (+/- 4)
NEW:
test iter::bench_trusted_random_access_chunks                      ... bench:          30 ns/iter (+/- 0)
```

The resulting assembly is similar to #103166 but the specialization kicks in under different (partially overlapping) conditions compared to that PR. They're complementary.

In principle a TRA-based specialization could be applied to all `ArrayChunks` methods, including `next()` as we do for `Zip` but that would have all the same hazards as the Zip specialization. Only doing it for `fold` is far less hazardous. The downside is that it only helps with internal, exhaustive iteration. I.e. `for _ in` or `try_fold` will not benefit.

Note that the regular, `try_fold`-based and the specialized `fold()` impl have observably slightly different behavior. Namely the specialized variant does not fetch the remainder elements from the underlying iterator. We do have a few other places in the standard library where beyond-the-end-of-iteration side-effects are being elided under some circumstances but not others.

Inspired by https://old.reddit.com/r/rust/comments/yaft60/zerocost_iterator_abstractionsnot_so_zerocost/

the8472 committed:
add benchmark for iter::ArrayChunks::fold specialization

This also updates the existing iter::Copied::next_chunk benchmark so
that the thing it benches doesn't get masked by the ArrayChunks specialization

the8472 committed:
make the array initialization guard available to other modules

the8472 committed:
specialize iter::ArrayChunks::fold for TrustedRandomAccess iters

This is fairly safe use of TRA since it consumes the iterator so
no struct in an unsafe state will be left exposed to user code

rust-highfive commented:
r? @Mark-Simulacrum

(rust-highfive has picked a reviewer for you, use r? to override)

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3307964613/jobs/5459995382) [(plain)](https://github.com/rust-lang/rust/commit/834078e740b31a724e2381c34d5041a63176e96a/checks/9056619054/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
Attempting with retry: make prepare
```

</details>

the8472 committed:
simplification: do not process the ArrayChunks remainder in fold()

Mark-Simulacrum commented on library/core/src/iter/adapters/array_chunks.rs:
```diff
@@ -168,3 +176,62 @@ where
+                // SAFETY: The method consumes the iterator and the loop condition ensures that
+                // all accesses are in bounds and only happen once.
+                guard.array_mut[j].write(unsafe { self.iter.__iterator_get_unchecked(i + j) });
+                guard.initialized = j + 1;
```
I guess this is fine, but is there a reason not to increment here? I would mildly expect that to be easier to make produce efficient code, though LLVM hopefully sees through most of this anyway.

the8472 replied:
No specific reason, I just haven't tried. Will do that.

the8472 replied:
replaced it with a loop using `guard.initialized` as induction variable.

Mark-Simulacrum commented on library/core/src/array/mod.rs:
```diff
@@ -916,6 +898,24 @@ where
+    fn drop(&mut self) {
+        debug_assert!(self.initialized <= N);
+
+        // SAFETY: this slice will contain only initialized objects.
```
Seems like this is a pretty lightly upheld condition, given the initialized is public (to the whole crate). Not very happy with that. But I'm not sure we have good options here...

the8472 replied:
I could add a bunch of unsafe methods instead. If only we had `pub(unsafe)` for field accesses...

Mark-Simulacrum replied:
I guess this seems like an ArrayVec abstraction anyway, just without all the APIs associated? Do we have traction on that (or perhaps have an impl there already)?

Mark-Simulacrum replied:
I'd be happier with an increment() unsafe method which seems like the only operation we need in the new code at least.

the8472 replied:
> Do we have traction on that (or perhaps have an impl there already)?

There's [a draft RFC](https://github.com/rust-lang/rfcs/pull/3316) but I think that's stuck in the "yes it would be great to have one, but before we do that there's this other allocator topic to research" hell.

We do have some building blocks of an arrayvec floating around, but not the whole thing. And one of the building blocks (`try_collect_into_array`) happens to not optimize well enough in some cases which is why I ended up handrolling it in more direct, unsafe ways in several PRs.

> I'd be happier with an increment() unsafe method which seems like the only operation

In principle touching any of the fields is unsafe, but ok, as long as we have something in the caller telling us that unsafe stuff happens that might be good enough.

Mark-Simulacrum replied:
Okay. It seems like internal ArrayVec shouldn't be controversial even if we don't expose it, but the optimizer point doesn't surprise me (though is also unfortunate).

r=me with increment added, I think it's fine for the time being, even though it's of course true that access to the array also ought to be unsafe.

the8472 replied:
Added documentation and an `unsafe fn push_unchecked()` that does more than just increment.

Mark-Simulacrum commented:
Seems broadly okay but I would like to hear your thoughts on especially the second comment around the safety condition.

the8472 committed:
document and improve array Guard type

The type is unsafe and now exposed to the whole crate.
Document it properly and add an unsafe method so the
caller can make it visible that something unsafe is happening.

the8472 commented:
@bors r=Mark-Simulacrum rollup

bors commented:
:pushpin: Commit 3925fc0c8e0b4ad9be48fbc1136281841d8985fa has been approved by `Mark-Simulacrum`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Mark-Simulacrum 3925fc0c8e0b4ad9be48fbc1136281841d8985fa -->
<!-- homu: {"type":"Approved","sha":"3925fc0c8e0b4ad9be48fbc1136281841d8985fa","approver":"Mark-Simulacrum","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

klensy commented:
Maybe don't rollup optimizations?

the8472 commented:
This is unstable API that's not used in the compiler yet, it should be fine.
